Source/ASDisplayNode.mm

```diff
@@ -2725,16 +2788,22 @@ - (void)_addSubnode:(ASDisplayNode *)subnode
 
 - (void)_addSubnodeViewsAndLayers
 {
-  for (ASDisplayNode *node in [_subnodes copy]) {
+  __instanceLock__.lock();
+  NSArray *subnodes = [_subnodes copy];
+  __instanceLock__.unlock();
+  for (ASDisplayNode *node in subnodes) {
     [self _addSubnodeSubviewOrSublayer:node];
   }
 }
 
 - (void)_addSubnodeSubviewOrSublayer:(ASDisplayNode *)subnode
 {
-    // Due to a bug in Apple's framework we have to use the layer index to insert a subview
-    // so just use th ecount of the sublayers to add the subnode
+  // Due to a bug in Apple's framework we have to use the layer index to insert a subview
+  // so just use th ecount of the sublayers to add the subnode
+  __instanceLock__.lock();
   NSInteger idx = _layer.sublayers.count;
+  __instanceLock__.unlock();
+  
```

Comment:
Since `_insertSubnodeSubviewOrSublayer` below is main thread only, we should assert main in this method and avoid locking. Same for `_addSubnodeViewsAndLayers`, `_loadViewOrLayerIsLayerBacked` and related methods.